### PR TITLE
New specs

### LINKED ISSUE #693: Future of Spec Strawman
"Yet another proposal for an alternative to Spec"

# The Status Quo
In D4J 2, there was no way for a user to build up a single http PATCH of multiple fields for an entity, forcing multiple http requests which could have been one. Because that's obviously inefficient, we wanted an elegant solution in D4J 3. The problem essentially decays into having some kind of named, default parameters in Java. From there, we realized that whatever solution we came up with could potentially be applied to other methods as well (even those not doing PATCH). This is traditionally solved with a Builder, but we didn't like forcing `new Builder()....build()` on common interactions with the user-facing api. Thus, [Spec was born](https://github.com/Discord4J/Discord4J/commit/e85ba3ed87051474188f327f7c4d3c500051f3fa).

Spec was just a mutable builder pattern without a finishing method (i.e., a builder without `build()`). Additionally, we realized that most people would simply construct these objects inline, so we provided `Consumer<Spec>` variants which would do that construction for you. Eventually, the "non-consumer variants" were [removed](https://github.com/Discord4J/Discord4J/commit/f92cf79a974dd29d5d8fca00e41eb8bed11544d7#diff-2c826afdc00bda17ee4e8944f741a496).

Specs have been somewhat of a pain point in D4J usage since 3.0. While they solved the aforementioned issues, they have many of their own. Namely,
1. They can still be verbose for simple things. (See #440)
2. They are mutable. (See #585)
3. Templating isn't very intuitive (See [Specs#Templates](https://github.com/Discord4J/Discord4J/wiki/Specs#templates))
4. Admittedly wrong usage can have weird behavior (See #646)

# Goals
To recap, the ideal solution...
1. Allows for a single request interacting with multiple properties
2. Is concise in both simple and complex cases
3. Plays nicely with Reactor
4. Can be templated intuitively

After seeing a [very similar solution](https://github.com/reactor/reactor-core/pull/1979) to a very similar problem in the community, we tentatively decided to move to immutable specs (#585). This solution achieves all but one goal: being concise. It is still very verbose to do simple things.

# Proposal
While working on transitioning to immutable specs, we came up with a new solution. At this point, we would like to receive community feedback. 

## API
```java
Mono<Role> editRole = role.edit()
  .withName("New Name")
  .withPermissions(PermissionSet.of(KICK_MEMBERS))
  .withReason("Because I want to");
```
Goal 1 is accomplished by the fact that this yields only one `Mono<Role>` representing the edit. ✔️ 

It is clearly concise, allowing for fluent chaining and not introducing new scopes. Goal 2 ✔️ 

It might be surprising that the type of this expression is actually `Mono<Role>`. How this is achieved will be exampled in Implementation. For now, note that this allows in-place chaining and direct integration with Reactor, accomplishing goal 3. ✔️ 

Finally, the template story:
```java
static RoleEditMono roleEditTemplate(Role role) {
  return role.edit().withColor(0x7289DA);
}
// ...
roleEditTemplate(role).withColor(override)
```
This could even potentially be altered to work in a `transform()` op to work better in a chain. Goal 4. ✔️ 

## Implementation
The API is accomplished by (potentially abusing?) implementing an "operator" on Mono. After [enabling copy methods](https://github.com/Discord4J/discord-json/commit/7eed123ef0f72ff0f1869a23a5abb8a8637f5a4f) on our json data classes, it is simple to make an immutable wrapper around that. Here is an example implementation:
```java
public class RoleEditMono extends Mono<Role> {

    private final GatewayDiscordClient gateway;
    private final long guildId;
    private final long roleId;

    private final ImmutableRoleModifyRequest request;
    @Nullable
    private final String reason;

    public RoleEditMono(GatewayDiscordClient gateway, long guildId, long roleId) {
        this(gateway, guildId, roleId, RoleModifyRequest.builder().build(), null);
    }

    // ... constructor ...

    public RoleEditMono withName(String name) {
        return withRequest(request.withName(name));
    }

    public RoleEditMono withPermissions(PermissionSet permissions) {
        return withRequest(request.withPermissions(permissions.getRawValue()));
    }

    // ... other properties ...

    private RoleEditMono withRequest(ImmutableRoleModifyRequest request) {
        return new RoleEditMono(this.gateway, this.guildId, this.roleId, request, this.reason);
    }

    @Override
    public void subscribe(CoreSubscriber<? super Role> actual) {
        gateway.rest().getGuildService()
                .modifyGuildRole(guildId, roleId, request, reason)
                .map(data -> new Role(gateway, data, guildId))
                .subscribe(actual);
    }
}
```
With a very straightforward implementation in `Role`:
```java
public RoleCreateMono edit() {
  return new RoleCreateMono(gateway, guildId, Snowflake.asLong(data.id()))
} 
```

As mentioned previously, we feel that this idea is ready for feedback. What are your comments, questions, concerns? 😄

### PR DESCRIPTION
**Pull request by darichey**
**Description:**
Initial work on new specs.

I've done a few already. If you want to help, make sure to follow the pattern. Feedback appreciated.

Some notes:
* I'm not necessarily satisfied with the attempt at abstraction/code sharing in `ConfigurableRequest` and `AuditableRequest`. I couldn't get the hierarchy to work well for things like `BanMono` which don't have internal builders. However, there are very few of these kinds of requests and it should be impossible for an outside user to trigger an error.
* AuditLogFlux also doesn't benefit from that code sharing, but it should eventually go away. See #428 

**Justification:**
Closes #693

### CODE REVIEW

File: core/src/main/java/discord4j/core/spec/GuildMemberEditMono.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        return withBuilder(it -> it.roles(list));
+    }
+
+    @Override
+    public GuildMemberEditMono withReason(String reason) {
+        return new GuildMemberEditMono(requestBuilder, reason, gateway, guildId, memberId);
+    }
+
+    @Override
+    Mono<Void> getRequest() {
```
Comment (quanticc) [inline core/src/main/java/discord4j/core/spec/GuildMemberEditMono.java:70]: In v8 this route returns MemberData

**Review by quanticc — CHANGES_REQUESTED** [review]

**Comment by quanticc** [comment]
Now I think we'll need to re-add the old Spec based methods. Deprecate them during 3.2 and removal should happen on 3.3

**Comment by quanticc** [comment]
@darichey I'm looking forward to merging this + #764 but maintaining old spec methods to provide a smoother migration path. What do you think?

**Comment by darichey** [comment]
Unfortunately, after a final push to get this done, I think I may have discovered something that could require rethinking our approach here. Consider the following method https://github.com/Discord4J/Discord4J/blob/884e0af6dd2e968dc3a576197103e4afe55d120a/core/src/test/java/discord4j/core/command/CommandOperations.java#L99-L105
How would this method be written with the proposed `MessageCreateMono`? I'm not sure it's possible. This was surprisingly the only method in our own codebase that does something like this, but I think it's a warning that this approach might not be all we thought it was.

Both this and the old approach take a kind of "ephemeral" approach to configuring actions. There is no concrete, tangible configuration data that can be manipulated like many people expect. Further, I think that's part of what leads to difficulty/confusion with the old design, and I'm not confident that this new design addresses that. Moreover, we may be further restricting the kinds of programs that can be written (assuming we cannot write the method above).

**Comment by quanticc** [comment]
I think it comes down whether we want to keep both approaches for v3.3. The example you quoted would have to use the current `createMessage(Consumer<? super MessageCreateSpec> spec)` right now, but it can be adapted if we accept externally configurable object to something like `fromSpec(...)` in this new PR.

**Comment by darichey** [comment]
I think ultimately this approach is going to cause problems. Unfortunately, I'm going to close this PR (we can keep the branch for a bit), and let's go back to immutable specs in #585.

**Pull request closed without merging**